---
title: "Nanopore scRNA-Seq Sicelore UMI vs wf-single-cell UMI benchmarks"
output:
  html_document:
    highlight: pygments
    df_print: paged
    toc: true
    toc_depth: 2
---

# Preparing the data

Load the required packages:
```{r, message = FALSE}
library(tidyverse)
library(glue)
library(scales)
library(SingleCellExperiment)
library(pheatmap)
library(viridis)
```

Global parameters:
```{r, message = FALSE, warning = FALSE}
result_dir <- file.path("..", "nanopore_sc_analysis", "report_data")
sample_ids <- c("SK-OV-3", "IGROV-1", "COV504")
min_tpm <- 1
min_cor_spearman <- 0.6
max_cor_spearman <- 0.9
heatmap_palette <- cividis
heatmap_palette_direction <- 1
```

Helper functions:
```{r, message = FALSE, warning = FALSE}
prepare_scatterplot_data <- function(tpm_matrix_1,
                                     tpm_matrix_2) {
    tpm_1_df <- as.data.frame(tpm_matrix_1)
    levels_1 <- colnames(tpm_1_df)
    tpm_1_df$transcript_id <- rownames(tpm_1_df)
    rownames(tpm_1_df) <- NULL
    tpm_1_df <- tpm_1_df %>%
        gather(-transcript_id, key = "sample_1", value = "tpm_1")
    tpm_1_df$sample_1 <- factor(
        tpm_1_df$sample_1, levels = levels_1
    )
    tpm_2_df <- as.data.frame(tpm_matrix_2)
    levels_2 <- colnames(tpm_2_df)
    tpm_2_df$transcript_id <- rownames(tpm_2_df)
    rownames(tpm_2_df) <- NULL
    tpm_2_df <- tpm_2_df %>%
        gather(-transcript_id, key = "sample_2",
               value = "tpm_2")
    tpm_2_df$sample_2 <- factor(
        tpm_2_df$sample_2, levels = levels_2
    )
    tpm_df <- left_join(tpm_1_df, tpm_2_df)
    tpm_df$status <- ifelse(
        as.numeric(tpm_df$sample_1) ==
            as.numeric(tpm_df$sample_2),
        "Correct vs Correct", "Correct vs Decoy"
    )
    tpm_df$status <- factor(
        tpm_df$status,
        levels = c("Correct vs Correct", "Correct vs Decoy")
    )
    return(tpm_df)
}
create_scatterplot_paired <- function(plot_df) {
    plot_df$tpm_1[plot_df$tpm_1 < 0.1] <- 0.1
    plot_df$tpm_2[plot_df$tpm_2 < 0.1] <- 0.1
    scatter_plot <- ggplot(plot_df, mapping = aes(x = tpm_1,
                                                  y = tpm_2)) +
        geom_point(alpha = 0.35, size = 0.05) +
        geom_abline(intercept = 0, slope = 1,
                    lty = 3, color = "red") +
        facet_grid(rows = vars(sample_2),
                   cols = vars(sample_1)) +
        scale_x_log10(labels = trans_format("log10", math_format(10^.x))) +
        scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
        coord_cartesian(xlim = c(0.1, 10^5),
                        ylim = c(0.1, 10^5),
                        clip = "off") +
        annotation_logticks(outside = TRUE,
                            size = 0.25,
                            short = unit(0.05, "cm"),
                            mid = unit(0.1, "cm"),
                            long = unit(0.15, "cm")) +
        theme_bw() +
        theme(aspect.ratio = 1,
              axis.title.x = element_blank(),
              axis.title.y = element_blank(),
              strip.text = element_text(size = 7))
    return(scatter_plot)
}
create_scatterplot_combined <- function(plot_df) {
    plot_df$tpm_1[plot_df$tpm_1 < 0.1] <- 0.1
    plot_df$tpm_2[plot_df$tpm_2 < 0.1] <- 0.1
    scatter_plot <- ggplot(plot_df, mapping = aes(x = tpm_1,
                                                  y = tpm_2)) +
        geom_point(alpha = 0.35, size = 0.05) +
        geom_abline(intercept = 0, slope = 1,
                    lty = 3, color = "red") +
        facet_grid(cols = vars(status)) +
        scale_x_log10(labels = trans_format("log10", math_format(10^.x))) +
        scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
        coord_cartesian(xlim = c(0.1, 10^5),
                        ylim = c(0.1, 10^5),
                        clip = "off") +
        annotation_logticks(outside = TRUE,
                            size = 0.25,
                            short = unit(0.05, "cm"),
                            mid = unit(0.1, "cm"),
                            long = unit(0.15, "cm")) +
        theme_bw() +
        theme(aspect.ratio = 1,
              axis.title.x = element_blank(),
              axis.title.y = element_blank())
    return(scatter_plot)
}
```

Prepare mitochondrial transcript blacklist:
```{r, message = FALSE, warning = FALSE}
se_transcript_mt <- readRDS(file.path(
    result_dir, "isosceles_se_transcript.rds"
))
se_transcript_mt <- se_transcript_mt[
    grepl("^MT:", rowData(se_transcript_mt)$position),
]
mt_transcripts_isosceles <- rownames(se_transcript_mt)
mt_transcripts_ensembl <- rowData(se_transcript_mt)$compatible_tx
# Manually removed outlier transcripts:
# ENST00000445125 (ISOT-0000-0000-0000-2622:s10197250:e10199150:FA:FL),
# RNA, 18S ribosomal pseudogene
# ENST00000536684 (ISOT-0000-0000-0000-caeb:s10507850:e10509200:FA:FL),
# MT-RNR2 like 8 (MTRNR2L8-201)
# ENST00000600213 (ISOT-0000-0000-0000-2809:s96617150:e96618250:FA:FL),
# MT-RNR2 like 12 (MTRNR2L12-201)
mt_transcripts_isosceles <- c(
    mt_transcripts_isosceles,
    c("ISOT-0000-0000-0000-2622:s10197250:e10199150:FA:FL",
      "ISOT-0000-0000-0000-caeb:s10507850:e10509200:FA:FL",
      "ISOT-0000-0000-0000-2809:s96617150:e96618250:FA:FL")
)
mt_transcripts_ensembl <- c(
    mt_transcripts_ensembl,
    c("ENST00000445125", "ENST00000536684", "ENST00000600213")
)
```

Prepare Isosceles pseudobulk RNA-Seq data (Sicelore input):
```{r, message = FALSE, warning = FALSE}
se_sicelore <- readRDS(file.path(
    result_dir, "isosceles_se_pseudobulk_transcript.rds"
))
se_sicelore <- se_sicelore[, 1:3]
colnames(se_sicelore) <- glue("{sample_ids} (Sicelore)")
tpm_sicelore <- assay(se_sicelore, "tpm")
tpm_sicelore <- tpm_sicelore[
    !(rownames(tpm_sicelore) %in% mt_transcripts_isosceles),
]
tpm_sicelore <- t(
    t(tpm_sicelore) / colSums(tpm_sicelore) * 1e6
)
```

Prepare Isosceles pseudobulk RNA-Seq data (wf-single-cell input):
```{r, message = FALSE, warning = FALSE}
se_wf <- readRDS(file.path(
    result_dir, "isosceles_wf_se_pseudobulk_transcript.rds"
))
se_wf <- se_wf[, 1:3]
colnames(se_wf) <- glue("{sample_ids} (wf-single-cell)")
tpm_wf <- assay(se_wf, "tpm")
tpm_wf <- tpm_wf[
    !(rownames(tpm_wf) %in% mt_transcripts_isosceles),
]
tpm_wf <- t(
    t(tpm_wf) / colSums(tpm_wf) * 1e6
)
```

Filter the results by minimal expression:
```{r, message = FALSE, warning = FALSE}
selector_sicelore <- apply(tpm_sicelore, 1, mean) >= min_tpm
selector_wf <- apply(tpm_wf, 1, mean) >= min_tpm
selector_expression <- selector_sicelore & selector_wf
```

Re-normalize the TPM values:
```{r, message = FALSE, warning = FALSE}
tpm_sicelore <- tpm_sicelore[selector_expression,]
tpm_wf <- tpm_wf[selector_expression,]
tpm_sicelore <- t(
    t(tpm_sicelore) / colSums(tpm_sicelore) * 1e6
)
tpm_wf <- t(
    t(tpm_wf) / colSums(tpm_wf) * 1e6
)
```

# Results

Spearman correlation matrix:
```{r, message = FALSE, warning = FALSE}
cor_spearman <- cor(tpm_wf, tpm_sicelore, method = "spearman")
as.data.frame(cor_spearman)
```

Spearman correlation heatmap:
```{r, message = FALSE, warning = FALSE}
pheatmap(
    cor_spearman,
    color = heatmap_palette(100, direction = heatmap_palette_direction),
    breaks = seq(min_cor_spearman, max_cor_spearman, length.out = 101),
    cluster_rows = FALSE, cluster_cols = FALSE, fontsize = 8
)
```

TPM scatter plots:
```{r, message = FALSE, warning = FALSE}
scatterplot_df <- prepare_scatterplot_data(tpm_sicelore, tpm_wf)
create_scatterplot_paired(scatterplot_df)
create_scatterplot_combined(scatterplot_df)
```

# Session Info

```{r}
sessionInfo()
```
