---
title: "Benchmarking Isosceles on SIRV data"
output:
  html_document:
    highlight: pygments
    df_print: paged
    toc: true
    toc_depth: 2
---

# Preparing the data

Load the required packages:
```{r, message = FALSE}
library(tidyverse)
library(glue)
library(pheatmap)
library(viridis)
library(SingleCellExperiment)
```

Global parameters:
```{r}
result_dir <- file.path("..", "sirv_analysis", "report_data")
sample_ids <- "cdna"
```

# Over-annotations analysis

Prepare reference data:
```{r, message = FALSE, warning = FALSE}
annotations_correct_df <- readRDS(file.path(
    result_dir, "annotations_correct.rds"
))$tx_df
correct_ref_tx_ids <- annotations_correct_df$compatible_tx %>%
    strsplit(",") %>%
    unlist() %>%
    sort()
```

Prepare Isosceles TPM data:
```{r, message = FALSE, warning = FALSE}
isosceles_tpm_list <- lapply(sample_ids, function(sample_id) {
    se_transcript <- readRDS(file.path(
        result_dir, glue("isosceles_{sample_id}_over_se_transcript.rds")
    ))
    tx_df <- as.data.frame(rowData(se_transcript))
    tpm_values <- assay(se_transcript, "tpm")[, 1]
    names(tpm_values) <- sapply(strsplit(tx_df$compatible_tx, ","), "[", 1)
    return(tpm_values)
})
names(isosceles_tpm_list) <- paste0("Isosceles.", sample_ids)
isosceles_correct_tpm_list <- lapply(isosceles_tpm_list, function(tpms) {
    tpms[names(tpms) %in% correct_ref_tx_ids]
})
isosceles_fractions <- sapply(isosceles_correct_tpm_list, sum) / sapply(isosceles_tpm_list, sum)
```

Prepare IsoQuant TPM data:
```{r, message = FALSE, warning = FALSE}
isoquant_tpm_list <- lapply(sample_ids, function(sample_id) {
    tx_df <- read_delim(file.path(
        result_dir, glue("isoquant_{sample_id}_over_transcript_tpm.tsv")
    ))
    tpm_values <- tx_df$TPM
    names(tpm_values) <- tx_df$`#feature_id`
    return(tpm_values)
})
names(isoquant_tpm_list) <- paste0("IsoQuant.", sample_ids)
isoquant_correct_tpm_list <- lapply(isoquant_tpm_list, function(tpms) {
    tpms[names(tpms) %in% correct_ref_tx_ids]
})
isoquant_fractions <- sapply(isoquant_correct_tpm_list, sum) / sapply(isoquant_tpm_list, sum)
```

Prepare bambu TPM data:
```{r, message = FALSE, warning = FALSE}
bambu_tpm_list <- lapply(sample_ids, function(sample_id) {
    tx_df <- read_delim(file.path(
        result_dir, glue("bambu_{sample_id}_over_counts_transcript.txt")
    ))
    tpm_values <- tx_df[, 3, drop = TRUE]
    tpm_values <- tpm_values / sum(tpm_values) * 1e6
    names(tpm_values) <- tx_df$TXNAME
    return(tpm_values)
})
names(bambu_tpm_list) <- paste0("bambu.", sample_ids)
bambu_correct_tpm_list <- lapply(bambu_tpm_list, function(tpms) {
    tpms[names(tpms) %in% correct_ref_tx_ids]
})
bambu_fractions <- sapply(bambu_correct_tpm_list, sum) / sapply(bambu_tpm_list, sum)
```

Plot fractions of TPMs from correct transcripts:
```{r, message = FALSE, warning = FALSE}
plot_df <- data.frame(
    tool = c("Isosceles", "IsoQuant", "Bambu"),
    value = c(
        isosceles_fractions["Isosceles.cdna"],
        isoquant_fractions["IsoQuant.cdna"],
        bambu_fractions["bambu.cdna"]
    )
)
plot_df$tool <- fct_reorder(plot_df$tool, plot_df$value)
ggplot(plot_df, mapping = aes(x = value,
                              y = tool,
                              fill = tool)) +
    geom_col() +
    geom_text(
        aes(label = format(round(value, 2), nsmall = 2)),
        size = 3,
        nudge_x = 0.03,
        hjust = "outward"
    ) +
    scale_fill_manual(values = c(Bambu = "mediumseagreen",
                                 Isosceles = "black",
                                 IsoQuant = "gold")) +
    lims(x = c(0, 1.1)) +
    labs(
        x = "Read precision (expression from correct transcripts)",
        y = ""
    ) +
    theme_classic() +
    theme(legend.position = "none",
          axis.text = element_text(size = 7),
          axis.title.y = element_blank(),
          axis.title.x = element_text(size = 8),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank())
```

# Insufficient annotations analysis

Prepare reference data:
```{r, message = FALSE, warning = FALSE}
annotations_correct_df <- readRDS(file.path(
    result_dir, "annotations_correct.rds"
))$tx_df
annotations_insufficient_df <- readRDS(file.path(
    result_dir, "annotations_insufficient.rds"
))$tx_df
correct_intron_positions <- annotations_correct_df$intron_positions
correct_intron_positions <- unique(correct_intron_positions[
    !is.na(correct_intron_positions)
])
insufficient_intron_positions <- annotations_insufficient_df$intron_positions
insufficient_intron_positions <- unique(insufficient_intron_positions[
    !is.na(insufficient_intron_positions)
])
```

Helper functions:
```{r, message = FALSE, warning = FALSE}
precision <- function(intron_positions) {
    TP <- sum(intron_positions %in% correct_intron_positions)
    FP <- sum(!intron_positions %in% correct_intron_positions)
    return(TP / (FP + TP))
}
recall <- function(intron_positions) {
    TP <- sum(intron_positions %in% correct_intron_positions)
    P <- length(correct_intron_positions)
    return(TP / P)
}
f1_score <- function(intron_positions) {
    TP <- sum(intron_positions %in% correct_intron_positions)
    FP <- sum(!intron_positions %in% correct_intron_positions)
    P <- length(correct_intron_positions)
    FN <- P - TP
    return(2 * TP / (2*TP + FP + FN))
}
classify_tx <- function(intron_positions) {
    return(c(
        TP = sum(intron_positions %in% correct_intron_positions),
        TP_KNOWN = sum(intron_positions %in% insufficient_intron_positions),
        TP_NOVEL = sum(
            (intron_positions %in% correct_intron_positions) &
                !(intron_positions %in% insufficient_intron_positions)
        ),
        FN = length(setdiff(correct_intron_positions, intron_positions)),
        FN_KNOWN = length(setdiff(
            insufficient_intron_positions, intron_positions
        )),
        FN_NOVEL = length(setdiff(
            setdiff(correct_intron_positions, insufficient_intron_positions),
            intron_positions
        )),
        FP = sum(!intron_positions %in% correct_intron_positions)
    ))
}
```

Prepare Isosceles intron position data:
```{r, message = FALSE, warning = FALSE}
isosceles_intron_list <- lapply(sample_ids, function(sample_id) {
    se_transcript <- readRDS(file.path(
        result_dir, glue("isosceles_{sample_id}_insufficient_se_transcript.rds")
    ))
    tpm_values <- assay(se_transcript, "tpm")[, 1]
    se_transcript <- se_transcript[tpm_values > 0,]
    intron_positions <- rowData(se_transcript)$intron_positions
    intron_positions <- unique(intron_positions[!is.na(intron_positions)])
    return(intron_positions)
})
names(isosceles_intron_list) <- paste0("Isosceles.", sample_ids)
```

Prepare IsoQuant intron position data:
```{r, message = FALSE, warning = FALSE}
isoquant_intron_list <- lapply(sample_ids, function(sample_id) {
    splicing_df <- Isosceles:::prepare_reference_annotations(file.path(
        result_dir, glue("isoquant_{sample_id}_insufficient_extended_annotations.gtf")
    ))$splicing_df
    tpm_df <- read_delim(file.path(
        result_dir, glue("isoquant_{sample_id}_insufficient_transcript_tpm.tsv")
    ))
    tpm_df <- tpm_df[tpm_df$TPM > 0,]
    expressed_tx_ids <- tpm_df[, 1, drop = TRUE]
    tpm_selector <- sapply(strsplit(splicing_df$compatible_tx, ","),
                           function(x) {any(x %in% expressed_tx_ids)})
    splicing_df <- splicing_df[tpm_selector,]
    intron_positions <- splicing_df$intron_positions
    return(intron_positions)
})
names(isoquant_intron_list) <- paste0("IsoQuant.", sample_ids)
```

Prepare bambu intron position data:
```{r, message = FALSE, warning = FALSE}
bambu_intron_list <- lapply(sample_ids, function(sample_id) {
    splicing_df <- Isosceles:::prepare_reference_annotations(file.path(
        result_dir, glue("bambu_{sample_id}_insufficient_extended_annotations.gtf")
    ))$splicing_df
    count_df <- read_delim(file.path(
        result_dir, glue("bambu_{sample_id}_insufficient_counts_transcript.txt")
    ))
    count_df <- count_df[count_df[, 3, drop = TRUE] > 0,]
    expressed_tx_ids <- count_df$TXNAME
    count_selector <- sapply(strsplit(splicing_df$compatible_tx, ","),
                             function(x) {any(x %in% expressed_tx_ids)})
    splicing_df <- splicing_df[count_selector,]
    intron_positions <- splicing_df$intron_positions
    return(intron_positions)
})
names(bambu_intron_list) <- paste0("bambu.", sample_ids)
```

Plot transcript detection metrics:
```{r, message = FALSE, warning = FALSE}
tools <- c("Isosceles", "IsoQuant", "Bambu")
plot_df <- data.frame(
    tool = rep(tools, each = 3),
    metric = rep(c("Precision", "Recall", "F1"), 3),
    value = c(
        precision(isosceles_intron_list[["Isosceles.cdna"]]),
        recall(isosceles_intron_list[["Isosceles.cdna"]]),
        f1_score(isosceles_intron_list[["Isosceles.cdna"]]),
        precision(isoquant_intron_list[["IsoQuant.cdna"]]),
        recall(isoquant_intron_list[["IsoQuant.cdna"]]),
        f1_score(isoquant_intron_list[["IsoQuant.cdna"]]),
        precision(bambu_intron_list[["bambu.cdna"]]),
        recall(bambu_intron_list[["bambu.cdna"]]),
        f1_score(bambu_intron_list[["bambu.cdna"]])
    )
)
plot_df$tool <- fct_reorder(plot_df$tool, plot_df$value)
plot_df$metric <- fct_rev(factor(
    plot_df$metric, levels = c("Precision", "Recall", "F1")
))
ggplot(plot_df, mapping = aes(x = value,
                              y = metric,
                              fill = tool)) +
    geom_col(position = position_dodge2()) +
    geom_text(
        aes(label = format(round(value, 2), nsmall = 2)),
        position = position_dodge2(width = .9),
        size = 3,
        hjust = -0.3
    ) +
    lims(x = c(0, 1.1)) +
    scale_fill_manual(values = c(Isosceles = "black",
                                 Bambu = "mediumseagreen",
                                 IsoQuant = "gold")) +
    theme_classic() +
    theme(legend.title = element_blank(),
          legend.position = "top",
          legend.key.size = unit(0.3, "cm"),
          legend.text = element_text(size = 7),
          axis.text = element_text(size = 7),
          axis.title.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank())
```

Transcript detection results:
```{r, message = FALSE, warning = FALSE}
classification_df <- data.frame(
    Isosceles = classify_tx(isosceles_intron_list[["Isosceles.cdna"]]),
    IsoQuant = classify_tx(isoquant_intron_list[["IsoQuant.cdna"]]),
    Bambu = classify_tx(bambu_intron_list[["bambu.cdna"]])
)
classification_df
```

Plot transcript detection results:
```{r, message = FALSE, warning = FALSE}
plot_df <- classification_df
plot_df$metric <- rownames(plot_df)
plot_df <- plot_df %>%
    pivot_longer(-metric, names_to = "tool", values_to = "value") %>%
    filter(metric %in% c("TP_KNOWN", "TP_NOVEL", "FP", "FN_KNOWN", "FN_NOVEL"))
plot_df$metric <- fct_rev(factor(
    plot_df$metric,
    levels = c(c("TP_KNOWN", "TP_NOVEL", "FN_KNOWN", "FN_NOVEL", "FP"))
))
plot_df$tool <- paste0(
    plot_df$tool, " ", ifelse(grepl("^TP", plot_df$metric), "TP", "FP/FN")
)
ggplot(plot_df, mapping = aes(x = value,
                              y = tool,
                              fill = metric)) +
    geom_col() +
    scale_fill_manual(values = c(TP_KNOWN = "darkgreen",
                                 TP_NOVEL = "mediumseagreen",
                                 FN_KNOWN = "royalblue4",
                                 FN_NOVEL = "cornflowerblue",
                                 FP = "darkred")) +
    theme_classic() +
    theme(legend.title = element_blank(),
          legend.position = "top",
          legend.key.size = unit(0.3, "cm"),
          legend.text = element_text(size = 5),
          axis.text = element_text(size = 7),
          axis.title.x = element_blank(),
          axis.title.y = element_blank())
```

# Session Info

```{r}
sessionInfo()
```
